storage: add limits to skipped data iteration

aliher1911 opened the pull request:
Previously when iterating engine using  MVCCIncrementalIterator caller
could skip large amounts of non-matching data which would result in
"unbounded" resource usage.
This is becoming a problem for resource constrained clusters where low
priority requests like export that are used by backups to interfere with
high priority workloads. If we want to throttle backups we need to be able
to limit how many underlying operations we want to perform per request.
This change adds an optional iteration limit to the iterator. Once the
limit is reached, iterator will return an error. Error will provide a
resume key to continue iteration in next request.

Release note: None

Fixes #68234

aliher1911 committed:
storage: add limits to skipped data iteration

Previously when exporting data from pebble iterator could spend unbouded
time skipping entries regardless of export size limits.
This is becoming a problem for resource constrained clusters where low
priority requests like export that are used by backups to interfere with
high priority workloads. If we want to throttle backups we need to be able
to limit how many underlying operations we want to perform per request.
This change adds an optional iteration limit to export. Once the
limit is reached, export will end its current chunk and return a resume
span even if desired size is not reached.
Current limiter uses wall clock time to stop interation.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68467)
<!-- Reviewable:end -->

aliher1911 commented:
This PR needs #68102 to land first because it would only work once we have an ability to stop on arbitrary key mid export.

I will update this one once it is ready. Regardless of its readiness, this functionality is distinct and could be reviewed independently and I don't want to put it in the stack of changes in previous PR.

aliher1911 committed:
storage: refactor export request params into struct

Number of arguments to ExportMVCCToSst is too large. This diff moves
them into a struct to improve readability.

Release note: None

Release justification:

aliher1911 commented:
@dt David do you think having an iteration limit should be a tenant/caller responsibility or a cluster responsibility? We have similar limit maxIntentCount which limits how many intents could be collected during export to have some protection for memory overuse and it is defined as a cluster setting because it protects servers from oom-ing. For iteration limits that throttle export resource usage on different dimension, should we stick to the same approach or rather let caller pass the limit explicitly? IIRC we wanted to give cpu constrained clients an ability to spread export requests, but it looks like it is more of a cluster level rather than request level limit.
I added it as export method argument, but having second thoughts before drilling it through export request api.

sumeerbhola reviewed (changes requested):
> Reviewed 5 of 10 files at r2, 5 of 5 files at r4, 1 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @dt, and @sumeerbhola)

___
*[pkg/storage/engine.go, line 362 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGiFbH5NSCHn7UyB8l:-MhGiFbH5NSCHn7UyB8m:b-d9r3n5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e84b4914b5ac490f410cb4a8f6eaea14bd70e3bf/pkg/storage/engine.go#L362)):*
> ```Go
> // ExportOptions contains options provided to export operation.
> type ExportOptions struct {
> 	// StartKey determines the start of exported interval (inclusive).
> ```

nit: ... start of the exported ...
___
*[pkg/storage/engine.go, line 363 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGiJe3126hMfGbKDQO:-MhGiJe3126hMfGbKDQP:byogdp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e84b4914b5ac490f410cb4a8f6eaea14bd70e3bf/pkg/storage/engine.go#L363)):*
> ```Go
> type ExportOptions struct {
> 	// StartKey determines the start of exported interval (inclusive).
> 	// StartKey.Timestamp is either empty which represent starting from potential
> ```

nit: represents starting from a potential ...
___
*[pkg/storage/engine.go, line 365 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGiT5cBNRFLO5ttAAX:-MhGiT5cBNRFLO5ttAAY:b-p6dfmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e84b4914b5ac490f410cb4a8f6eaea14bd70e3bf/pkg/storage/engine.go#L365)):*
> ```Go
> 	// StartKey.Timestamp is either empty which represent starting from potential
> 	// intent and continuing to versions or non-empty, which represents starting
> 	// from particular version.
> ```

nit: from a particular version.
___
*[pkg/storage/engine.go, line 367 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGiVkTAPUSuUejnyKC:-MhGiVkU66Dsl9n0YCbO:bbmojxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e84b4914b5ac490f410cb4a8f6eaea14bd70e3bf/pkg/storage/engine.go#L367)):*
> ```Go
> 	// from particular version.
> 	StartKey MVCCKey
> 	// EndKey determines end of exported interval (exclusive).
> ```

nit: determines the end ...
___
*[pkg/storage/engine.go, line 369 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGiglnAfwHG4bP83TH:-MhGiglnAfwHG4bP83TI:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e84b4914b5ac490f410cb4a8f6eaea14bd70e3bf/pkg/storage/engine.go#L369)):*
> ```Go
> 	// EndKey determines end of exported interval (exclusive).
> 	EndKey roachpb.Key
> 	// StartTS and EndTS determine exported time range as (startTS, endTS]
> ```

nit: missing period.
___
*[pkg/storage/engine.go, line 387 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGivf_9j6fZPydItG5:-MhGivf_9j6fZPydItG6:b-lt7ybp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e84b4914b5ac490f410cb4a8f6eaea14bd70e3bf/pkg/storage/engine.go#L387)):*
> ```Go
> 	// If StopMidKey is false, once function reaches targetSize it would continue
> 	// adding all versions until it reaches next key or end of range. If true, it
> 	// would stop immediately when targetSize is reached and return a next versions
> ```

nit: ... the next versions ...
___
*[pkg/storage/mvcc_incremental_iterator.go, line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGlq1b14MbopWZLep9:-MhGlq1b14MbopWZLepA:b-c48u82) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee6a9d5bcca9a1edddcbfb71fdcedaf905a15e8/pkg/storage/mvcc_incremental_iterator.go#L130)):*
> ```Go
> 	// Resume key is not necessarily a valid iteration key as we could stop in between
> 	// eligible keys.
> 	MaxAllowedIterations int64
> ```

I'm wary of introducing resource control settings that cannot be easily understood in terms of real resources.
maxIntentCount was not quite a real resource but could be understood as a limit on the return size.

Ideally, we would want to limit the cpu time and IO time spent in executing an operation that scans data. This would also fit in well with the current admission control which functions better when the execution of a `BatchRequest` has a bounded size. But we don't have cpu information from golang and I can't see when we we would have IO time information either.

But I think this could still be something more tangible that we can use not just for exports, but also for other scans and gets (requests for non-background operations could use a higher limit). We have some counts in `pebble.IteratorStats` but they have multiple dimensions like step/seek, forward/reverse and it would be nicer to have a single dimension that was more relatable. I am thinking number of ssblock bytes "read" (we'd count all the bytes for an ssblock when loading it into the iterator) would be a good metric. If the setting puts a 100MB limit on it, it means something real. We don't expose this value via `IteratorStats`, or track it in the low-level sstable iterators but it can be added. One downside is that it does not count the bytes iterated in the memtable -- I don't think that matters in production settings at all since the memtable is tiny relative to the rest of the store.
@jbowens @itsbilal for other opinions.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @itsbilal, @jbowens, and @sumeerbhola)

___
*[pkg/storage/mvcc_incremental_iterator.go, line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGlq1b14MbopWZLep9:-MhIq3r1BOksyYjOpGHJ:b-ivyju4) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee6a9d5bcca9a1edddcbfb71fdcedaf905a15e8/pkg/storage/mvcc_incremental_iterator.go#L130)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm wary of introducing resource control settings that cannot be easily understood in terms of real resources.
maxIntentCount was not quite a real resource but could be understood as a limit on the return size.

Ideally, we would want to limit the cpu time and IO time spent in executing an operation that scans data. This would also fit in well with the current admission control which functions better when the execution of a `BatchRequest` has a bounded size. But we don't have cpu information from golang and I can't see when we we would have IO time information either.

But I think this could still be something more tangible that we can use not just for exports, but also for other scans and gets (requests for non-background operations could use a higher limit). We have some counts in `pebble.IteratorStats` but they have multiple dimensions like step/seek, forward/reverse and it would be nicer to have a single dimension that was more relatable. I am thinking number of ssblock bytes "read" (we'd count all the bytes for an ssblock when loading it into the iterator) would be a good metric. If the setting puts a 100MB limit on it, it means something real. We don't expose this value via `IteratorStats`, or track it in the low-level sstable iterators but it can be added. One downside is that it does not count the bytes iterated in the memtable -- I don't think that matters in production settings at all since the memtable is tiny relative to the rest of the store.
@jbowens @itsbilal for other opinions.
</blockquote></details>

I think like the idea of limiting on block-bytes loaded, but as far as a KV-level work limit, I think capping the number of keys compared for inclusion in the returned result seems reasonable; we already have limits on number of keys returned by KV, so limiting number of keys that we will examined by KV _for_ return, even if they aren't ultimately returned, doesn't seem like too much of a conceptual stretch? 

Ultimately I think both limits would be desirable -- limiting block bytes loaded helps cap the IO and storage CPU footprint of the request, but if for example we're in some very compressed, valueless index-key blocks, I could see wanting to limit the number of iterations, not just loaded block bytes, to also limit KV's cpu time too. 


<!-- Sent from Reviewable.io -->

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @itsbilal, @jbowens, and @sumeerbhola)

___
*[pkg/storage/mvcc_incremental_iterator.go, line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGlq1b14MbopWZLep9:-MhJObBX7Ih44_4L0jJT:b39d53t) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee6a9d5bcca9a1edddcbfb71fdcedaf905a15e8/pkg/storage/mvcc_incremental_iterator.go#L130)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think like the idea of limiting on block-bytes loaded, but as far as a KV-level work limit, I think capping the number of keys compared for inclusion in the returned result seems reasonable; we already have limits on number of keys returned by KV, so limiting number of keys that we will examined by KV _for_ return, even if they aren't ultimately returned, doesn't seem like too much of a conceptual stretch? 

Ultimately I think both limits would be desirable -- limiting block bytes loaded helps cap the IO and storage CPU footprint of the request, but if for example we're in some very compressed, valueless index-key blocks, I could see wanting to limit the number of iterations, not just loaded block bytes, to also limit KV's cpu time too. 
</blockquote></details>

My 2p on what I see our goal here.
We want to limit resources used by export requests to minimize impact on higher priority requests. The most obvious resources are CPU and IO in this case, but also a time that we held range locks. I'm not sure what is the memory impact at the moment. Depending on range content we could have many small values that we need to go through which could incur extra CPU load when skipping data or large payloads where we can hit higher IO use.

We also need a way to expose that to MVCCIncrementalIterator which is currently agnostic of underlying storage and delegates work to storage iterators.
Current approach with just counting how many times we stepped avoid adding extra methods to iterators, but it is a bit ugly and only acting as a proxy of underlying complexity.

A better approach may be to get a "counters" object when constructing iterator in reader.NewMVCCIterator() and then just ask that object if we reached target?

The second aspect that I don't particularly like is that we explicitly pass limits which couples caller with iterator while a better way could be to have some reading "profile" that would be configured separately and could have as many resource limits as we currently have so that we don't have to change the whole call stack if we find a way to expose something more useful.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @sumeerbhola)

___
*[pkg/storage/mvcc_incremental_iterator.go, line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGlq1b14MbopWZLep9:-Mhmx9Q57F4Ea-QaZ_Cm:b-u5qvzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee6a9d5bcca9a1edddcbfb71fdcedaf905a15e8/pkg/storage/mvcc_incremental_iterator.go#L130)):*
> The most obvious resources are CPU and IO in this case, but also a time that we held range locks.

How about we measure walltime. Unless the system is very overloaded, it should be a good proxy for cpu+io time.

 > The second aspect that I don't particularly like is that we explicitly pass limits which couples caller with iterator while a better way could be to have some reading "profile" that would be configured separately and could have as many resource limits as we currently have so that we don't have to change the whole call stack if we find a way to expose something more useful.

This seems worthwhile. We could add something like a `ResourceTimeLimiter` struct, which for now would only do walltime, and pass it through `MVCCIncrementalIteratOptions`/`MVCCScanOptions`/`MVCCGetOptions` (like we do with the `mon.BoundAccount` in the latter two). This would be called every N loop iterations to amortize the cost of fetching the "resource time", to check if some resource time had been exceeded. Later we could add other dimensions to the `ResourceTimeLimiter`.


<!-- Sent from Reviewable.io -->

aliher1911 committed:
kvserver: add cluster configuration to throttle exports

Export requests could iterate unbounded amount of data in storage
this diff adds kv.bulk_sst.max_request_time hidden cluster setting
to limit how for long export could run irrespective of how much data
is actually exported.

Release note: None

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @sumeerbhola)

___
*[pkg/storage/mvcc_incremental_iterator.go, line 130 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MhGlq1b14MbopWZLep9:-MhnyKmY3vW068kaunKL:bpjj7fp) ([raw file](https://github.com/cockroachdb/cockroach/blob/eee6a9d5bcca9a1edddcbfb71fdcedaf905a15e8/pkg/storage/mvcc_incremental_iterator.go#L130)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> The most obvious resources are CPU and IO in this case, but also a time that we held range locks.

How about we measure walltime. Unless the system is very overloaded, it should be a good proxy for cpu+io time.

 > The second aspect that I don't particularly like is that we explicitly pass limits which couples caller with iterator while a better way could be to have some reading "profile" that would be configured separately and could have as many resource limits as we currently have so that we don't have to change the whole call stack if we find a way to expose something more useful.

This seems worthwhile. We could add something like a `ResourceTimeLimiter` struct, which for now would only do walltime, and pass it through `MVCCIncrementalIteratOptions`/`MVCCScanOptions`/`MVCCGetOptions` (like we do with the `mon.BoundAccount` in the latter two). This would be called every N loop iterations to amortize the cost of fetching the "resource time", to check if some resource time had been exceeded. Later we could add other dimensions to the `ResourceTimeLimiter`.
</blockquote></details>

I gave it a go and it looks ok I think. There's still an issue with ensuring that we always advance, but it is currently mitigated by throttling checks so that we always move N times before first check is done.
Maybe we could pass current key to Exhausted() so that it could check or have this check in iterator itself which would require saving start key and comparing current to start once we hit the limit.
If there would be other checks, limiter might need a view of underlying iterator stats injected or the whole creation sequence turned upside down.


<!-- Sent from Reviewable.io -->

sumeerbhola commented:
@aliher1911 I've been distracted with other things. Are you trying to merge this for 21.2?

aliher1911 commented:
@sumeerbhola I think it is too late for 21.1.0 now, but its been dragging too long so I would appreciate some feedback so I can finalize it soon.

sumeerbhola reviewed (changes requested):
> Reviewed 3 of 11 files at r10, 2 of 7 files at r12, 1 of 532 files at r14, 2 of 529 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @dt, @itsbilal, @jbowens, and @sumeerbhola)

___
*[pkg/storage/mvcc_incremental_iterator.go, line 41 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbAs9bCVBTfGN7ub3J:-MjbAs9bCVBTfGN7ub3K:b-e640ga) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L41)):*
> ```Go
> //
> // If iteration resource limit is requested, iterator would return an error as
> // soon as limit is reached. The error will contain a resume key which could be
> ```

why would this be outside the requested span?
___
*[pkg/storage/mvcc_incremental_iterator.go, line 132 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbBZC_1x-zfdX6EE78:-MjbBZC_1x-zfdX6EE79:bhiz5mu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L132)):*
> ```Go
> // ClockCheckEveryNIterations defines for how many iterations we could cache
> // current time when performing iteration wall clock time limiting.
> const ClockCheckEveryNIterations = 100
> ```

can this be package private?
___
*[pkg/storage/mvcc_incremental_iterator.go, line 240 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbFsl43qYFo5H3E0zU:-MjbFsl43qYFo5H3E0zV:bwz0p7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L240)):*
> ```Go
> 	// StartKey must also be populated with resume key. This is needed to ensure progress
> 	// for cases when initial seek would exhaust resources and that subsequent call would
> 	// restart from further position.
> ```

I didn't understand this comment. Is this trying to say that we don't want to stop until we are past `startKey`? Given this is a `SimpleMVCCIterator` the first call on this needs to be `SeekGE`. We can keep a `notFirstCallToSeekGE` bool in `MVCCIncrementalIterator` and and pass it to `advance()`.

Something like
```
func (...) SeekGE(...) {
   ...
   ignoreLimiter := !i.notFirstCallToSeekGE
   i.notFirstCallToSeekGE = false
   i.advance(ignoreLimiter)
}
```
I think this is better than doing additional key comparisons, and doesn't require additional state in options.
 
___
*[pkg/storage/mvcc_incremental_iterator.go, line 242 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbBtTP1eH93T5byg9o:-MjbBtTP1eH93T5byg9p:b-uzh34i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L242)):*
> ```Go
> 	// restart from further position.
> 	// Note that resume key is not necessarily a valid iteration key as we could stop in
> 	// between eligible keys because of timestamp range limits.
> ```

because of resource limits.


<!-- Sent from Reviewable.io -->

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @itsbilal, @jbowens, and @sumeerbhola)

___
*[pkg/storage/mvcc_incremental_iterator.go, line 41 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbAs9bCVBTfGN7ub3J:-Mje_VbpANRVVrwOLcGF:b-i50ffi) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L41)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why would this be outside the requested span?
</blockquote></details>

If we don't use TBI then when we do Next on underlying iter it may move to the mvcckey that is out of bounds for iterator. We then do advance which will start iterating until it hits next valid mvcckey, but resource limiter could cut it short so we end up with the key that is not eligible for inclusion into results.
With TBI it should always be within limits I think.
___
*[pkg/storage/mvcc_incremental_iterator.go, line 132 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbBZC_1x-zfdX6EE78:-MjeaZnK0bbrLpBoZFD3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L132)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can this be package private?
</blockquote></details>

Done.
___
*[pkg/storage/mvcc_incremental_iterator.go, line 240 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbFsl43qYFo5H3E0zU:-MjeacSCEpjZzitsCcge:bortxg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L240)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I didn't understand this comment. Is this trying to say that we don't want to stop until we are past `startKey`? Given this is a `SimpleMVCCIterator` the first call on this needs to be `SeekGE`. We can keep a `notFirstCallToSeekGE` bool in `MVCCIncrementalIterator` and and pass it to `advance()`.

Something like
```
func (...) SeekGE(...) {
   ...
   ignoreLimiter := !i.notFirstCallToSeekGE
   i.notFirstCallToSeekGE = false
   i.advance(ignoreLimiter)
}
```
I think this is better than doing additional key comparisons, and doesn't require additional state in options.
 
</blockquote></details>

I like that, it looks cleaner at least on the outside.
___
*[pkg/storage/mvcc_incremental_iterator.go, line 242 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbBtTP1eH93T5byg9o:-MjebLBqD1U6YbRKlMkM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L242)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

because of resource limits.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

aliher1911 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @itsbilal, @jbowens, and @sumeerbhola)

___
*[pkg/storage/mvcc_incremental_iterator.go, line 41 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbAs9bCVBTfGN7ub3J:-MjhvWLa74Ew0YVa_vHX:b-30hght) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L41)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

If we don't use TBI then when we do Next on underlying iter it may move to the mvcckey that is out of bounds for iterator. We then do advance which will start iterating until it hits next valid mvcckey, but resource limiter could cut it short so we end up with the key that is not eligible for inclusion into results.
With TBI it should always be within limits I think.
</blockquote></details>

I was thinking about it and it could be fixed by adding checks before error is raised and adjusting the key in the following way - if timestamp is lower than requested, move to the next key and use latest requested timestamp, if timestamp is higher, move to the highest requested without changing key, if the key moved outside of requested range as a result of adjustment, don't err and finish gracefully.
That would duplicate the logic we have within the loop. Do you think this added complexity is justified for consistency?


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68467)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @dt, @itsbilal, @jbowens, and @sumeerbhola)

___
*[pkg/storage/mvcc_incremental_iterator.go, line 41 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/68467#-MjbAs9bCVBTfGN7ub3J:-Mjo34zD8Pbf92H5s4LT:bag4ruv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e92b5a2dd21eff3a87d5a8ff5b27db67ed71e/pkg/storage/mvcc_incremental_iterator.go#L41)):*
> If we don't use TBI then when we do Next on underlying iter it may move to the mvcckey that is out of bounds for iterator.

We set an upper bound for the iter, yes? So I am not sure why that would happen.

> We then do advance which will start iterating until it hits next valid mvcckey, but resource limiter could cut it short so we end up with the key that is not eligible for inclusion into results.

But I think I agree that the general structure of this change is error prone. The `MVCCIncrementalIterator` is maintaining some invariants that are true at the end of each seek and next* function. By checking for resource limits in the middle of these functions after some work is done, we are inviting complexity regarding those invariants. An easy fix would be lift the resource limit checking to the start of each seek and next* function and avoid doing the checking for the first seek, to ensure forward progress.
That brings me to a related concern: the code in pebble.go isn't necessarily trivial since it has to deal with the fact that `MVCCIncrementalIterator` can complain about resource limits at any call to next*. There are 2 additional users of `MVCCIncrementalIterator` in `catchup_scan.go and mvcc.go. The former is probably not one that will use resource limits, but the latter eventually could. Can you look at the code to see if it would be simpler to lift the resource limit checking into the callers? With the current approach we can't avoid modifying the callsites, so why not give them full control. That way if they don't want the resource checking to happen before some particular next call since it is harder to maintain some resumption invariant, they can do so.

Let me know if you want to discuss synchronously.


<!-- Sent from Reviewable.io -->

aliher1911 commented:
Summarizing: we had an offline conversation with @sumeerbhola and he made fair points that time bound iterator should reduce severity of effects when we need to skip over large amounts of data. Without such requirement, code that tracks resources could be pulled out on level above so that export itself could check how long we iterate and stop if needed.
This should be good enough until we see any evidence that we are really blocked by iterator making slow Next() calls.

Second consideration is that mvcc_incremental_iterator maintains some invariants regarding its state and what it returns and breaking out of loop violates them. Having such loosely defined component is not good and it would impact planned work to rewrite it in near future.

Based on that discussion I made changes to pull code out to pebble part.

sumeerbhola commented:
Thanks for summarizing. Adding to the considerations in the previous comment: the clients of `MVCCIncrementalIterator` are also not always able to stop and resume at arbitrary points in the iteration. e.g. if `stopMidKey` is false in `pebbleExportToSst`. By lifting the logic up into the client, they can take into account their own stopping constraints.

sumeerbhola commented on pkg/storage/engine.go:
```diff
@@ -387,6 +387,13 @@ type ExportOptions struct {
+	// of it once resources are exhausted.
+	// This option could only be used together with StopMidKey as it could break
+	// iteration over data on any version and even on a key and timestamp that would
+	// not be included in the resulting export.
```
Is this comment up to date? I would expect we don't need to limit this to StopMidKey now that the decision on when to stop is not in the MVCCIterator code.

sumeerbhola commented on pkg/storage/resource_limiter.go:
```diff
@@ -0,0 +1,76 @@
+const (
+	ResourceLimitNotReached  ResourceLimitReached = 0
+	ResourceLimitReachedSoft ResourceLimitReached = 1
+	ResourceLimitReachedHard ResourceLimitReached = 2
```
Why do we need the soft and hard limits? If we simply have a limit, once the caller knows the limit is exceeded it knows it needs to stop as soon as viable (given the callers internal invariants).

aliher1911 replied:
This is in line with what we do for sstable size. Callers would prefer to have chunks split by key, in that case they can process them immediately. But if we hit hard limit we stop asap and then caller will have to deal with only a part of versions for the key and combine them with the following chunk.

sumeerbhola commented on pkg/storage/resource_limiter_test.go:
```diff
@@ -0,0 +1,84 @@
+				break
+			}
+			it.Next()
+			sumTime += timeutil.Since(startClock).Nanoseconds()
```
What was the benchmark difference with and without this call to `timeutil.Since`?

aliher1911 replied:
Sorry for the confusion, this test should go completely. I was just using the branch to push benchmark to the linux server to check the overheads.

sumeerbhola commented on pkg/storage/pebble.go:
```diff
@@ -2034,10 +2034,30 @@ func pebbleExportToSst(
 			curKey = append(curKey[:0], unsafeKey.Key...)
 		}
 
+		if options.ResourceLimiter != nil {
```
I like how much simpler this is. And if we remove the difference between hard and soft limits (other comment) it will be even simpler.

aliher1911 replied:
To reiterate, export requests would prefer to get sstables split on key boundaries to avoid additional processing in tenant. So having two thresholds doesn't seem excessively complex.

sumeerbhola reviewed (commented):
Generally looks good. A few minor comments

sumeerbhola reviewed (approved):
lgtm

aliher1911 commented:
bors r=sumeerbhola
